packages/@aws-cdk/aws-servicecatalog/test/product.test.ts

```diff
@@ -271,94 +271,26 @@ describe('Product', () => {
-        productName: 'testProduct',
-        owner: 'testOwner',
-        productVersions: [
-          {
-            cloudFormationTemplate: servicecatalog.CloudFormationTemplate.fromUrl('https://awsdocs.s3.amazonaws.com/servicecatalog/development-environment.template'),
-          },
-        ],
-      });
-    }),
-
-    test('add tag options to product', () => {
-      const tagOptions = new servicecatalog.TagOptions({
-        key1: ['value1', 'value2'],
-        key2: ['value1'],
-      });
-
-      product.associateTagOptions(tagOptions);
-
-      Template.fromStack(stack).resourceCountIs('AWS::ServiceCatalog::TagOption', 3); //Generates a resource for each unique key-value pair
-      Template.fromStack(stack).resourceCountIs('AWS::ServiceCatalog::TagOptionAssociation', 3);
-    }),
-
-    test('add tag options as input to product in props', () => {
-      const tagOptions = new servicecatalog.TagOptions({
-        key1: ['value1', 'value2'],
-        key2: ['value1'],
-      });
-
-      new servicecatalog.CloudFormationProduct(stack, 'MyProductWithTagOptions', {
-        productName: 'testProduct',
-        owner: 'testOwner',
-        productVersions: [
-          {
-            cloudFormationTemplate: servicecatalog.CloudFormationTemplate.fromUrl('https://awsdocs.s3.amazonaws.com/servicecatalog/development-environment.template'),
-          },
-        ],
-        tagOptions: tagOptions,
-      });
-
-      Template.fromStack(stack).resourceCountIs('AWS::ServiceCatalog::TagOption', 3); //Generates a resource for each unique key-value pair
-      Template.fromStack(stack).resourceCountIs('AWS::ServiceCatalog::TagOptionAssociation', 3);
-    }),
-
-    test('adding identical tag options to product is idempotent', () => {
-      const tagOptions1 = new servicecatalog.TagOptions({
-        key1: ['value1', 'value2'],
-        key2: ['value1'],
-      });
-
-      const tagOptions2 = new servicecatalog.TagOptions({
-        key1: ['value1', 'value2'],
-      });
-
-      product.associateTagOptions(tagOptions1);
-      product.associateTagOptions(tagOptions2); // If not idempotent this would fail
-
-      Template.fromStack(stack).resourceCountIs('AWS::ServiceCatalog::TagOption', 3); //Generates a resource for each unique key-value pair
-      Template.fromStack(stack).resourceCountIs('AWS::ServiceCatalog::TagOptionAssociation', 3);
-    }),
-
-    test('adding duplicate tag options to portfolio and product creates unique tag options and enumerated associations', () => {
-      const tagOptions1 = new servicecatalog.TagOptions({
-        key1: ['value1', 'value2'],
-        key2: ['value1'],
-      });
-
-      const tagOptions2 = new servicecatalog.TagOptions({
-        key1: ['value1', 'value2'],
-        key2: ['value2'],
-      });
-
-      const portfolio = new servicecatalog.Portfolio(stack, 'MyPortfolio', {
-        displayName: 'testPortfolio',
-        providerName: 'testProvider',
-      });
+  });
```

Comment:
okay, that makes sense.  I'll remove changes to the individual unit test files and just have the added tests here.  In separate PR we can remove the redundancies.

Reply:
👍

Reply:
As a follow up, I think we need to change the functionality and some of these tests, I should've explicitly put this in breaking changes.

For example, I don't think the idempotent associations were that meaningful, and maintaining it now becomes difficult since we would have to scan entire construct tree for the app to find if there is similar key-value, since they could be in different `TagOptions` constructs, I don't believe we can use `tryFindChild`.  I will continue to look for way around this, since this breaks current functionality.

Reply:
We could solve problem partially by storing the key and value alongside the CfnTagOptions.  And checking if those exist.

Reply:
And again the tests that have input validation are now not meaningful, previously it would break when calling the `associate` method.  But now it will fail to create the `TagOptions` themselves, thus no need to test it in the portfolio unit tests.

Reply:
Let's just remove the tests that don't make sense anymore.

Reply:
> But now it will fail to create the `TagOptions` themselves, thus no need to test it in the portfolio unit tests.

It doesn't matter where it fails. Like I said, we shouldn't be moving the tests methods at the same time we are changing production code.

So, please leave the validation tests where they are.

Reply:
We can move them later, in a separate PR, that doesn't touch production code, only tests.